### PR TITLE
Add script to include HTML snippets

### PR DESCRIPTION
**Pull request by Angel-AG**
There used to be some repeated content (`<header>`, `<nav>`, `<footer>`) across different files.
These contents were separated in individual files, to allow future changes to be made in a single file.
The HTML snippets are included through a HTTP Request.

### CODE REVIEW

File: portfolio/src/main/webapp/reusehtml.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    const path = element.getAttribute("include-html");
+    
+    // Check path is not an empty string
+    if (path) {
+      // Make an HTTP Request for the HTML file
+      const xmlHttp = new XMLHttpRequest();
+      xmlHttp.onreadystatechange = function() {
+        if (xmlHttp.readyState === 4) {
+          if (xmlHttp.status === 200) {element.innerHTML = xmlHttp.responseText;};
+          if (xmlHttp.status === 404) {element.innerHTML = "Page not found";};
```
Comment (mschillaci) [inline portfolio/src/main/webapp/reusehtml.js]: Could these be an if and else-if, and then you could even have a fallback else case for when the status isn't something you expect. What happens if you get a 403 for example, both these are skipped but the element.removeAttribute is still run?

File: portfolio/src/main/webapp/reusehtml.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  for (const element of elementsWithIncludeAttr) {
+    // Get the file name
+    const path = element.getAttribute("include-html");
+    
+    // Check path is not an empty string
+    if (path) {
+      // Make an HTTP Request for the HTML file
+      const xmlHttp = new XMLHttpRequest();
+      xmlHttp.onreadystatechange = function() {
+        if (xmlHttp.readyState === 4) {
```
Comment (mschillaci) [inline portfolio/src/main/webapp/reusehtml.js]: I'm not sure how many states there can be, but perhaps a fallback here as well. Are there error states, what if it hangs, is there a timeout on the request?

**Review by mschillaci — COMMENTED** [review]
Looks good, made a few comments, mainly around error cases and edge cases. Everything looks fine for what we would call "happy path", but it's a good exercise to strengthen all the not-happy path cases as well.

File: portfolio/src/main/webapp/reusehtml.js (hunk 1 above)
Comment (Angel-AG) [reply portfolio/src/main/webapp/reusehtml.js]: Thanks! Makes sense to have a fallback else case. I added also another else-if case for timeout.
I am leaving the element.removeAttribute there, it runs when the HTTP Request is done, even if the response was unfortunate, and the attribute is not going to be used again after the request is done, so it is removed.

File: portfolio/src/main/webapp/reusehtml.js (hunk 2 above)
Comment (Angel-AG) [reply portfolio/src/main/webapp/reusehtml.js]: There are 5, from 0 to 4. They monitor the progress, with 4 indicating that the operation is complete. I include a loading message while the state is not 4.
I also added a timeout of 5 seconds, when it is reached for the first time it calls a function to try the request again, if it fails again it shows a timeout message.
To handle errors, xmlHttp.onerror was defined.

File: portfolio/src/main/webapp/reusehtml.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  xmlHttp.ontimeout = timeoutFunc;
+
+  xmlHttp.send();
+}
+
+/**
+ * Adds HTML content to elements with the include-html attribute
+ */
+function includeHtml() {
+  const elementsWithIncludeAttr = document.querySelectorAll("[include-html]");
```
Comment (akihiroota87) [inline portfolio/src/main/webapp/reusehtml.js]:
```suggestion
  const elements = document.querySelectorAll("[include-html]");
```
Nit: I think a simpler name is fine here.

File: portfolio/src/main/webapp/reusehtml.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+  for (const element of elementsWithIncludeAttr) {
+    // Get the file name
+    const path = element.getAttribute("include-html");
+    
+    // Check path is not an empty string
+    if (path) {
+      // Two tries to request the file
```
Comment (akihiroota87) [inline portfolio/src/main/webapp/reusehtml.js:83]: Why do we try twice? How was this number decided? Is there any reason the first try wouldn't succeed?

**Review by akihiroota87 — APPROVED** [review]
Awesome looks good once comments are addressed!

File: portfolio/src/main/webapp/reusehtml.js (hunk 3 above)
Comment (Angel-AG) [reply portfolio/src/main/webapp/reusehtml.js]: Okay! Will do!

File: portfolio/src/main/webapp/reusehtml.js (hunk 4 above)
Comment (Angel-AG) [reply portfolio/src/main/webapp/reusehtml.js:83]: When running the project in the testing environment, there are times when some parts of the page do not load (like images or scripts, or even the whole page). I am not sure if this is something related to the testing environment, but sometimes I was experiencing that a request did not move forward and stay pending, so I arbitrarily decide to give it a "refresh" when the timeout is reached.

**Review by mschillaci — APPROVED** [review]
Looks good to me!

File: portfolio/src/main/webapp/reusehtml.js (hunk 4 above)
Comment (akihiroota87) [reply portfolio/src/main/webapp/reusehtml.js:83]: Okay, thank you for the clarification. That is very strange, maybe we can take a closer look at the issue in our 1:1.

Comment (Angel-AG) [reply portfolio/src/main/webapp/reusehtml.js:83]: Yep, sure.

**Pull request merged**